ipa-core/src/helpers/buffers/circular.rs

```diff
@@ -0,0 +1,788 @@
+/// the capacity limit, because it can be set large enough.
+///
+/// This buffer can also be closed, using [`close`] method. After it is
+/// closed, it allows reads of any size, ubt it guarantees that all of them
+/// will be aligned with `write_size`.
+///
+/// ## Implementation notes
+/// This buffer is built over a [`Vec`] with two extra pointers that indicate
+/// the place to read from, and to write next chunk to. When read happens,
+/// it moves the read pointer until it meets the write pointer.
+/// When read points to the same location as write, this buffer is considered
+/// empty.
+///
+/// Both pointers operate within the range `[0, 2*capacity)` and clamped into
+/// the working range when used as index into the internal buffer. The reason
+/// is to be able to distinguish empty and full buffers.
+///
+/// This implementation does not perform checks in optimized builds,
+/// relying on [`OrderingSender`] to enforce correctness. If taken away,
+/// necessary adjustments need to be made to avoid data corruption.
+///
+/// ## Alternative implementations
+/// If alignment to `read_size` is too much, a [`BipBuffer`] can be used instead.
+///
+/// ## Future improvements
+/// [`OrderingSender`] currently synchronizes reader and writers, but it does not
+/// have to if this implementation is made thread-safe. There exists a well-known
+/// lock-free FIFO implementation for a single producer, single consumer that uses
+/// atomics for read and write pointers. We can't make use of it as is because there
+/// are more than one writer. However, [`OrderingSender`] already knows now to allow
+/// only one write at a time, so it could be possible to make the entire
+/// implementation lock-free.
+///
+/// [`BipBuffer`]: <https://www.codeproject.com/Articles/3479/The-Bip-Buffer-The-Circular-Buffer-with-a-Twist>
+/// [`OrderingSender`]: crate::helpers::buffers::OrderingSender
+/// [`can_read`]: CircularBuf::can_read
+/// [`close`]: CircularBuf::close
+pub struct CircularBuf {
+    /// write pointer, points to the beginning of the next write slice
+    write: usize,
+    /// read pointer, points to the beginning of the next read slice
+    read: usize,
+    /// The size of read returned from [`take`] if buffer is not closed and not empty.
+    read_size: usize,
+    /// The size of one write
+    write_size: usize,
+    /// Whether this buffer is closed
+    closed: bool,
+    /// Actual data, stored inside a contiguous region in memory.
+    data: Vec<u8>,
+}
+
+impl CircularBuf {
+    /// Constructs a new instance of [`CircularBuf`] with reserved `capacity` bytes and specified
+    /// `write_size` and `read_size` bytes.
+    ///
+    /// ## Panics
+    /// If any of the following conditions are met:
+    /// * Any provided value is 0
+    /// * `write_size` is not a multiple of `capacity`
+    /// * `read_size` is not a multiple of `write_size`
+    /// * `read_size` is smaller than `write_size`
```

Comment:
Are you sure? I thought that was the whole idea, that we want small reads and big writes. Why would it panic?

Reply:
Also, I don't see this enforced by any asserts below.

Reply:
it is enforced by `read_size % write_size == 0` check and validated by `panic_on_bad_read_size` test

`write_size` defines the size of a single write - usually just a few bytes, the size of a single record. In the vectorized case, this may be one kibibyte but still less than or equal to the size of a single read from this buffer. We don't want fragmentation here, so a single record should be entirely in one batch sent over the network